src/Algebra/Module/Construct/Nagata.agda

```diff
@@ -0,0 +1,170 @@
+------------------------------------------------------------------------
+-- The Agda standard library
+--
+-- The non-commutative analogue of Nagata's construction
+-- of the "idealization of a module", defined here
+-- on R-R-*bi*modules M over a ring R.
+--
+-- Elements of R ⋉ M are pairs |R| × |M|
+------------------------------------------------------------------------
+
+{-# OPTIONS --cubical-compatible --safe #-}
+
+module Algebra.Module.Construct.Nagata where
+
+open import Algebra.Bundles
+open import Algebra.Core
+import Algebra.Consequences.Setoid as Consequences
+import Algebra.Definitions as Definitions
+open import Algebra.Module.Bundles
+open import Algebra.Module.Core
+open import Algebra.Module.Definitions
+import Algebra.Module.Construct.DirectProduct as DirectProduct
+import Algebra.Module.Construct.TensorUnit as TensorUnit
+open import Algebra.Structures
+open import Data.Product using (_×_; _,_; proj₁; proj₂)
+open import Data.Product.Relation.Binary.Pointwise.NonDependent
+open import Level using (Level; _⊔_)
+open import Relation.Binary using (Rel; Setoid; IsEquivalence)
+import Relation.Binary.Reasoning.Setoid as ≈-Reasoning
+
+private
+  variable
+    m ℓm r ℓr : Level
+
+------------------------------------------------------------------------
+-- Definitions
+
+module Nagata (ring : Ring r ℓr) (bimodule : Bimodule ring ring m ℓm) where
+
+  open Ring ring
+    renaming (Carrier to R
+             ; 0# to 0ᴿ
+             ; 1# to 1ᴿ
+             ; _+_ to _+ᴿ_
+             ; _*_ to _*ᴿ_
+             ; *-cong to *ᴿ-cong
+             ; *-assoc to *ᴿ-assoc
+             ; *-identityʳ to *ᴿ-identityʳ
+             ; *-identityˡ to *ᴿ-identityˡ
+             ; distribˡ to *ᴿ-distribˡ-+ᴿ
+             ; distribʳ to *ᴿ-distribʳ-+ᴿ
+             )
+
+  open Bimodule bimodule
+    renaming (Carrierᴹ to M
+             )
+
+  open AbelianGroup +ᴹ-abelianGroup
+    renaming (setoid to setoidᴹ
+             ; isEquivalence to isEquivalenceᴹ
+             )
+
+  open IsEquivalence isEquivalenceᴹ
+    renaming (sym to symᴹ)
+
+  +ᴺ-bimodule = DirectProduct.bimodule TensorUnit.bimodule bimodule
+
+  open Bimodule +ᴺ-bimodule public using ()
+    renaming (Carrierᴹ to N
+             ; _≈ᴹ_ to _≈ᴺ_
+             ; _+ᴹ_ to _+ᴺ_
+             ; 0ᴹ to 0ᴺ
+             ; -ᴹ_ to -ᴺ_
+             ; +ᴹ-abelianGroup to +ᴺ-abelianGroup
+             )
+
+  ιᴿ : R → N
+  ιᴿ r = r , 0ᴹ
+
+  ιᴹ : M → N
+  ιᴹ m = 0ᴿ , m
+
+  1ᴺ : N
+  1ᴺ = ιᴿ 1ᴿ
+
+  open AbelianGroup +ᴺ-abelianGroup
+    renaming (isAbelianGroup to +ᴺ-isAbelianGroup)
+
+  open Definitions _≈ᴺ_
+
+  infixl 7 _*ᴺ_
+
+  _*ᴺ_ : Op₂ N
+  (r₁ , m₁) *ᴺ (r₂ , m₂) = r₁ *ᴿ r₂ , r₁ *ₗ m₂ +ᴹ m₁ *ᵣ r₂
+
+  *ᴺ-identity : Identity 1ᴺ _*ᴺ_
+  *ᴺ-identity = lᴺ , rᴺ
+    where
+      open ≈-Reasoning setoidᴹ
+      lᴺ : LeftIdentity 1ᴺ _*ᴺ_
```

Comment:
Can we lift the left and right proofs (here and in distributivity) up to the top-level please? They're often useful to have individually.

Reply:
Ah!? OK, but I assume they would be available as projections once the `IsRing` and `Ring` instances were defined (and opened suitably... ;-)), ...

... but will do. What about naming though, if they are to be publicly visible?

Reply:
So... attempting to standardise names so that the `(Is)Ring` structure/bundle for `N` used the 'usual' names, I got a strange constraint error when trying to `open Definitions _≈_` after rename `_≈ᴹ_` to `_≈_`, so I left that one as-is.